### PR TITLE
store dataset as a sparse scipy matrix

### LINKED ISSUE #57: np sparse
As discussed, default format should be sparse matrices, and we densify only when sending on the gpu

### PR DESCRIPTION
**Pull request by maxime1310**
The dataset class now handles both numpy arrays or scipy CSC/CSR sparse matrix. If the data is stored as a sparse matrix, the data will be densified on the fly as the get_item method or get_all method is called.

We'll check this afternoon whether using this sparse format instead of densifiying all the matrix in the beginning is faster.

closes #57

### CODE REVIEW

File: run_benchmarks.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
 if __name__ == '__main__':
     parser = argparse.ArgumentParser()
     parser.add_argument("--epochs", type=int, default=250, help="how many times to process the dataset")
-    parser.add_argument("--dataset", type=str, default="cortex", help="which dataset to process")
+    parser.add_argument("--dataset", type=str, default="brain_large", help="which dataset to process")
```
Comment (jeff-regier) [inline run_benchmarks.py]: Probably good to keep `cortex` the default since `brain_large` is too large to run on many computers.

Comment (jeff-regier) [reply run_benchmarks.py]: Oh, I see you're subsampling. Still, it takes awhile to download.

File: scvi/dataset/brain_large.py [hunk 2]
```diff
@@ -72,7 +72,8 @@ def __init__(self, subsample_size=20):
         self.genome = "mm10"
         self.download_and_preprocess()
         h5_object = get_matrix_from_h5(self.save_path + self.final_name, self.genome)
-        super(BrainLargeDataset, self).__init__([np.array(h5_object.matrix.transpose().todense())])
+        super(BrainLargeDataset, self).__init__([h5_object.matrix.transpose().tocsr(copy=False)], sparse=True)
+        # super(BrainLargeDataset, self).__init__([h5_object.matrix.transpose()], sparse=True)
```
Comment (jeff-regier) [inline scvi/dataset/brain_large.py]: Better not to commit commented out code.

File: scvi/dataset/dataset.py [hunk 3]
```diff
@@ -13,32 +14,39 @@ class GeneExpressionDataset(Dataset):
...
 
-    def __init__(self, Xs, list_labels=None):
+    def __init__(self, Xs, list_labels=None, sparse=False):
         # Args:
         # Xs: a list of numpy tensors with .shape[1] identical (total_size*nb_genes)
+        # or a list of scipy CSR sparse matrix,
+        # or transposed CSC sparse matrix (the argument sparse must then be set to true)
         self.nb_genes = Xs[0].shape[1]
         self.n_batches = len(Xs)
+        self.sparse = sparse
```
Comment (jeff-regier) [inline scvi/dataset/dataset.py]: Do we ever want `Xs` not be be sparse? I think yesterday @romain-lopez said all the rna-seq data is sparse.

File: scvi/dataset/dataset.py [hunk 4]
```diff
@@ -49,13 +57,23 @@ def __init__(self, Xs, list_labels=None):
             self.all_labels = torch.zeros_like(self.batch_indices)  # We might want default label values
 
     def get_all(self):
```
Comment (jeff-regier) [inline scvi/dataset/dataset.py]: Can we delete `get_all`? It doesn't look like it's called anywhere. I don't think its something we'd ever want to use---we should always visit the dataset as minibatches.

File: run_benchmarks.py (hunk 1 above)
Comment (maxime1310) [reply run_benchmarks.py]: I agree with you, I just put that back to Cortex.

File: scvi/dataset/dataset.py [hunk 5]
```diff
@@ -13,32 +14,39 @@ class GeneExpressionDataset(Dataset):
...
         self.n_batches = len(Xs)
+        self.sparse = sparse
+
         assert all(X.shape[1] == self.nb_genes for X in Xs), "All tensors must have same size"
 
         new_Xs = []
         local_means = []
         local_vars = []
         batch_indices = []
         for i, X in enumerate(Xs):
```
Comment (jeff-regier) [inline scvi/dataset/dataset.py]: Loops in Python are very slow and therefore aren't idiomatic. You almost never want to iterate over all the rows of an array. Can you do things like `Xs.mean(0)` (row mean) and `Xs.mean(1)` (column mean) here instead?

Comment (Edouard360) [reply scvi/dataset/dataset.py]: @jeff-regier What for loop are you referring to ?

https://github.com/YosefLab/scVI-dev/blob/cd91fac1fb0edff59e5189e155aaa9e9a43230ed/scvi/dataset/dataset.py#L24-L35

In the above, Xs refers to a list of different (biological) batch (to take into account batch effects), and so we separate the computation of the l_mu, and l_var for each of these batch. Did you mean that or something else ?

**Comment by Edouard360** [comment]
solves #57

**Comment by Edouard360** [comment]
resolves #57

File: run_benchmarks.py [hunk 6]
```diff
@@ -10,8 +10,8 @@
 
 if __name__ == '__main__':
     parser = argparse.ArgumentParser()
-    parser.add_argument("--epochs", type=int, default=250, help="how many times to process the dataset")
-    parser.add_argument("--dataset", type=str, default="cortex", help="which dataset to process")
+    parser.add_argument("--epochs", type=int, default=25, help="how many times to process the dataset")
```
Comment (jeff-regier) [inline run_benchmarks.py]: Is this change just for testing? You can do it in pycharm wo changing the code, by specifying different command line arguments in the run configurations.

File: scvi/dataset/dataset.py [hunk 7]
```diff
@@ -56,24 +54,13 @@ def __init__(self, Xs, list_labels=None, sparse=False):
...
     def __getitem__(self, idx):
-        if self.sparse:
-            return torch.Tensor(np.resize(self.X[idx, :].toarray(), (self.X[idx, :].toarray().shape[1]))), \
-                   self.local_means[idx], self.local_vars[idx], self.batch_indices[idx], self.all_labels[idx]
-        else:
-            return torch.Tensor(self.X[idx]),\
-                   self.local_means[idx], self.local_vars[idx], self.batch_indices[idx], self.all_labels[idx]
+
+        return torch.Tensor(np.resize(self.X[idx, :].toarray(), (self.X[idx, :].toarray().shape[1]))), \
+               self.local_means[idx], self.local_vars[idx], self.batch_indices[idx], self.all_labels[idx]
```
Comment (jeff-regier) [inline scvi/dataset/dataset.py]: Nice! Much cleaner now.

File: run_benchmarks.py (hunk 6 above)
Comment (Edouard360) [reply run_benchmarks.py]: Yes this is a mistake, I changed it back in last commit

**Pull request merged**